### PR TITLE
Planner: Support Planner Interpreter

### LINKED ISSUE #4739: interpreter planner refactor
#### background
Currently tiflash's interpreter is designed based on `DAGQueryBlock`.
`DAGQueryBlock` is designed to coprocess read, and it can handle simple queries well.
However, there is no way to handle the complex query of MPP, so the `DAGQueryBlock` tree is introduced, which is a compromise that makes the code more complicated and affects the subsequent iterative development.
We need to refactor the interpreter code in a plan visit way.

#### design
Therefore we introduce a new abstraction layer called `PhysicalPlan` to replace `DAGQueryBlock`.
It will be responsible for the processing of `ExpressionAction` and the generation of `BlockInputStream`.
We call this organization of code the planner.

```
DAGRequest (executor tree or executor array).
     |
     |
     v
physical plan <---+
     |            |
     |            | optimize
     |            |
     +------------+   
     |
     |
     v
BlockInputStream
```

#### why planner
Most databases are designed based on planner. 
- Friendly for database developers to understand code.
- We can improve tiflash with a general approach from the DB area, such as pipeline mode.

#### todo list
- [x] public desgin doc
    - #4743
- [x] Small refactoring of Interpreter before planner refactoring 
     - #4118 
- [x] interpreter test framework 
    - #4609 
- [x] planner refactory
    - [x] poc #4601 
    - [x] Introduce `PhysicalPlan`/`planner`, add switch and set it to planner by default
        - https://github.com/pingcap/tiflash/pull/4820 
    - [x] Replace all operators with `PhysicalPlan`
        - [x] common plans: projection/agg/filter/exchange_(sender/receiver)/limit/topn
            - https://github.com/pingcap/tiflash/pull/5143
        - [x] window function
            - https://github.com/pingcap/tiflash/pull/5363
        - [x] join
            - https://github.com/pingcap/tiflash/pull/5320
        - [x] table scan
            - https://github.com/pingcap/tiflash/pull/5335
    - [x] remove `DAGQueryBlock`
            - https://github.com/pingcap/tiflash/pull/5381
- [x] merge to master
    - feature branch: https://github.com/pingcap/tiflash/tree/planner_refactory
    - https://github.com/pingcap/tiflash/pull/5493
    - https://github.com/pingcap/tiflash/pull/5321

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #4739

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support Planner Interpreter.
```

### CODE REVIEW

**Comment by SeaRise** [comment]
wait for https://github.com/pingcap/tiflash/pull/5493

**Comment by SeaRise** [comment]
wait for https://github.com/pingcap/tiflash/pull/5493

File: .github/workflows/license-checker.yml [hunk 1]
```diff
@@ -3,10 +3,10 @@ name: License checker
 on:
   push:
     branches:
-      - master
+      - planner_refactory
   pull_request:
     branches:
-      - master
+      - planner_refactory
```
Comment (SeaRise) [inline .github/workflows/license-checker.yml:9]: After the protection of planner_refactory is turned off, I will restore the changes here.

File: dbms/src/Flash/Planner/ExecutorIdGenerator.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            return fmt::format("{}_aggregation", ++current_id);
+        case tipb::ExecType::TypeTopN:
+            return fmt::format("{}_top_n", ++current_id);
+        case tipb::ExecType::TypeLimit:
+            return fmt::format("{}_limit", ++current_id);
+        case tipb::ExecType::TypeExchangeSender:
+            return fmt::format("{}_exchange_sender", ++current_id);
+        case tipb::ExecType::TypeExchangeReceiver:
+            return fmt::format("{}_exchange_receiver", ++current_id);
+        default:
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/ExecutorIdGenerator.h:64]: window and sort must have executor_id.

Comment (SeaRise) [reply dbms/src/Flash/Planner/ExecutorIdGenerator.h:64]: But we can also consider adding processing to them, after merging master..

File: dbms/src/Flash/executeQuery.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#include <Interpreters/executeQuery.h>
+
+namespace DB
+{
+BlockIO executeQuery(
+    Context & context,
+    bool internal,
+    QueryProcessingStage::Enum stage)
+{
+    if (context.getSettingsRef().enable_planner)
```
Comment (SeaRise) [inline dbms/src/Flash/executeQuery.cpp:27]: The switch to planner works here.

File: dbms/src/Flash/tests/gtest_aggregation_executor.cpp [hunk 4]
```diff
@@ -163,6 +163,7 @@ class ExecutorAggTestRunner : public DB::tests::ExecutorTest
 TEST_F(ExecutorAggTestRunner, GroupBy)
 try
 {
+    WRAP_FOR_DIS_ENABLE_PLANNER_BEGIN
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_aggregation_executor.cpp:166]: Used to test enable planner and disable planner

File: dbms/src/Flash/Planner/PlanType.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+namespace DB
+{
+struct PlanType
+{
+    enum PlanTypeEnum
+    {
+        Limit = 0,
+        TopN = 1,
+        Filter = 2,
+        Aggregation = 3,
```
Comment (fuzhe1989) [inline dbms/src/Flash/Planner/PlanType.h:28]: (just discussion) if someone wants to split `Aggregation` into `PartialAggregation` and `FinalAggregation`, do you think it's good to change here?

Comment (fuzhe1989) [reply dbms/src/Flash/Planner/PlanType.h:28]: My question is what's the right way to insert `LocalShuffle` into a plan?

Comment (SeaRise) [reply dbms/src/Flash/Planner/PlanType.h:28]: > (just discussion) if someone wants to split `Aggregation` into `PartialAggregation` and `FinalAggregation`, do you think it's good to change here?

LGTM

Comment (SeaRise) [reply dbms/src/Flash/Planner/PlanType.h:28]: > (just discussion) if someone wants to split `Aggregation` into `PartialAggregation` and `FinalAggregation`, do you think it's good to change here?

> (just discussion) if someone wants to split `Aggregation` into `PartialAggregation` and `FinalAggregation`, do you think it's good to change here?

Here is an attempt at splitting agg.
https://github.com/pingcap/tiflash/pull/4874

**Comment by SeaRise** [comment]
> /label tide/merge-method-merge

This label will make this pr use non-squash-merge

File: dbms/src/Flash/Planner/PlanType.h (hunk 5 above)
Comment (SeaRise) [reply dbms/src/Flash/Planner/PlanType.h:28]: > My question is what's the right way to insert `LocalShuffle` into a plan?

In `PhysicalPlan::build` or `optimize`?

File: dbms/src/Flash/Planner/FinalizeHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        ", ");
+    bf.append("}");
+    return bf.toString();
+}
+} // namespace
+
+void prependProjectInputIfNeed(ExpressionActionsPtr & actions, size_t columns_from_previous)
+{
+    if (!actions->getRequiredColumnsWithTypes().empty()
+        && columns_from_previous > actions->getRequiredColumnsWithTypes().size())
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:65]: this is unlikely to happen.
We can add assert here...

File: dbms/src/Flash/Planner/FinalizeHelper.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <unordered_map>
+
+namespace DB::FinalizeHelper
+{
+namespace
+{
+String namesToString(const Names & names)
+{
+    return fmt::format("{{{}}}", fmt::join(names, ","));
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:32]: But it's really an array.

File: dbms/src/Flash/Planner/FinalizeHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+{
+    NameSet schema_set;
+    for (const auto & column : schema)
+        schema_set.insert(column.name);
+    for (const auto & parent_require_column : parent_require)
+    {
+        if (unlikely(schema_set.find(parent_require_column) == schema_set.end()))
+            throw TiFlashException(
+                fmt::format("schema {} don't contain parent require column: {}", schemaToString(schema), parent_require_column),
+                Errors::Planner::Internal);
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:81]: How about creating an issue to replace all `if (unlikely(condition)) { throw Exception }`

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:81]: Not only code for planner :)

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:81]: https://github.com/pingcap/tiflash/issues/5527

File: dbms/src/Flash/Planner/FinalizeHelper.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (fuzhe1989) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:32]: that's my point: array is usually wrapped by `[]` not `{}`.

File: dbms/src/Flash/Planner/FinalizeHelper.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (fuzhe1989) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:81]: great!

File: dbms/src/Flash/Coprocessor/PushDownFilter.h [hunk 9]
```diff
@@ -25,6 +25,10 @@ struct PushDownFilter
 {
     static PushDownFilter toPushDownFilter(const String & executor_id, const tipb::Executor * executor);
 
+    static PushDownFilter toPushDownFilter(const String & executor_id, const tipb::Selection & selection);
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/PushDownFilter.h:28]: `pushDownFilterFromExecutor` and `pushDownFilterFromSelection`?

File: dbms/src/Flash/Planner/FinalizeHelper.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/FinalizeHelper.cpp:32]: ok

File: dbms/src/Flash/Coprocessor/PushDownFilter.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (fuzhe1989) [reply dbms/src/Flash/Coprocessor/PushDownFilter.h:28]: the arguments already contain enough information. do not repeat yourself.

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/PushDownFilter.h:28]: okk, `pushDownFilterFrom`?

**Comment by SeaRise** [comment]
> > /label tide/merge-method-merge
> 
> This label will make this pr use non-squash-merge

Now decided to squash merge， and the label removed.

**Review by windtalker — APPROVED** [review]
LGTM

File: dbms/src/Flash/Planner/PlanType.cpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+// limitations under the License.
+
+#include <Common/TiFlashException.h>
+#include <Flash/Planner/PlanType.h>
+
+namespace DB
+{
+String PlanType::toString() const
+{
+    switch (enum_value)
```
Comment (ywqzzy) [inline dbms/src/Flash/Planner/PlanType.cpp:22]: Can use magic_enum lib in future.

File: dbms/src/Flash/Planner/FinalizeHelper.cpp [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+#include <Interpreters/ExpressionActions.h>
+#include <common/types.h>
+
+#include <unordered_map>
+
+namespace DB::FinalizeHelper
+{
+namespace
+{
+String namesToString(const Names & names)
```
Comment (ywqzzy) [inline dbms/src/Flash/Planner/FinalizeHelper.cpp:30]: maybe  namesToString,schemaToString, blockMetaToString should be in Names.h, NamesAndTypes.h and Block.h

File: dbms/src/Flash/Planner/plans/PhysicalJoin.cpp [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+void PhysicalJoin::transformImpl(DAGPipeline & pipeline, Context & context, size_t max_streams)
+{
+    /// The build side needs to be transformed first.
+    {
+        DAGPipeline build_pipeline;
+        build()->transform(build_pipeline, context, max_streams);
+        buildSideTransform(build_pipeline, context, max_streams);
+    }
+
+    {
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Planner/plans/PhysicalJoin.cpp:251]: Just wanted to make the distinction between probe and build a little clearer

File: dbms/src/Flash/Coprocessor/FineGrainedShuffle.h [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <Flash/Coprocessor/DAGContext.h>
+#include <common/types.h>
+#include <tipb/executor.pb.h>
+
+namespace DB
+{
+static constexpr std::string_view enableFineGrainedShuffleExtraInfo = "enable fine grained shuffle";
+
+inline bool enableFineGrainedShuffle(uint64_t stream_count)
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/FineGrainedShuffle.h:25]: This is to make DAGQueryBlockInterpreter compile without error.
- https://github.com/pingcap/tiflash/blob/45792ca39d5a84d5090c27db14844fd1788b5af0/dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp#L464
- https://github.com/pingcap/tiflash/blob/45792ca39d5a84d5090c27db14844fd1788b5af0/dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp#L619-L625

File: dbms/src/Flash/Coprocessor/PushDownFilter.cpp [hunk 14]
```diff
@@ -49,17 +49,22 @@ tipb::Executor * PushDownFilter::constructSelectionForRemoteRead(tipb::Executor
...
     }
 
+    return pushDownFilterFrom(executor_id, executor->selection());
+}
+
+PushDownFilter PushDownFilter::pushDownFilterFrom(const String & executor_id, const tipb::Selection & selection)
+{
     std::vector<const tipb::Expr *> conditions;
-    for (const auto & condition : executor->selection().conditions())
+    for (const auto & condition : selection.conditions())
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/PushDownFilter.cpp:65]: Using `selection.conditions()` will make a big change to the original code...
Not just `appendWhere()`, but `DAGQueryInfo` etc.

File: dbms/src/Core/ColumnsWithTypeAndName.h [hunk 15]
```diff
@@ -23,4 +23,5 @@ namespace DB
 {
 using ColumnsWithTypeAndName = std::vector<ColumnWithTypeAndName>;
 
+
```
Comment (ywqzzy) [inline dbms/src/Core/ColumnsWithTypeAndName.h:26]: nit: remove empty line.

Comment (SeaRise) [inline dbms/src/Core/ColumnsWithTypeAndName.h:26]:
```suggestion
```

File: dbms/src/Flash/Planner/plans/PhysicalBinary.h [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        , left(left_)
+        , right(right_)
+    {
+        RUNTIME_ASSERT(left, log, "children(0) shouldn't be nullptr");
+        RUNTIME_ASSERT(right, log, "children(1) shouldn't be nullptr");
+    }
+
+    PhysicalPlanNodePtr children(size_t i) const override
+    {
+        RUNTIME_ASSERT(i <= 1, log, "child_index({}) should not >= childrenSize({})", i, childrenSize());
```
Comment (ywqzzy) [inline dbms/src/Flash/Planner/plans/PhysicalBinary.h:46]:
```suggestion
        RUNTIME_ASSERT(i <= 1, log, "child_index({}) shouldn't >= childrenSize({})", i, childrenSize());
```

File: dbms/src/Flash/Planner/plans/PhysicalBinary.h [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        return i == 0 ? left : right;
+    }
+
+    void setChild(size_t i, const PhysicalPlanNodePtr & new_child) override
+    {
+        RUNTIME_ASSERT(i <= 1, log, "child_index({}) should not >= childrenSize({})", i, childrenSize());
```
Comment (ywqzzy) [inline dbms/src/Flash/Planner/plans/PhysicalBinary.h:52]:
```suggestion
        RUNTIME_ASSERT(i <= 1, log, "child_index({}) shouldn't >= childrenSize({})", i, childrenSize());
```

**Review by ywqzzy — APPROVED** [review]
LGTM

**Pull request merged**